include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h

```diff
@@ -66,36 +212,37 @@ struct __parallel_sort_submitter<_IdType, __internal::__optional_kernel_name<_Le
                                  __internal::__optional_kernel_name<_GlobalSortName...>,
                                  __internal::__optional_kernel_name<_CopyBackName...>>
 {
-    template <typename _BackendTag, typename _ExecutionPolicy, typename _Range, typename _Compare>
+    template <typename _BackendTag, typename _ExecutionPolicy, typename _Range, typename _Compare, typename _LeafSorter>
     auto
-    operator()(_BackendTag, _ExecutionPolicy&& __exec, _Range&& __rng, _Compare __comp) const
+    operator()(_BackendTag, _ExecutionPolicy&& __exec, _Range&& __rng, _Compare __comp,
+               _LeafSorter& __leaf_sorter) const
     {
         using _Tp = oneapi::dpl::__internal::__value_t<_Range>;
         using _Size = oneapi::dpl::__internal::__difference_t<_Range>;
 
         const std::size_t __n = __rng.size();
         assert(__n > 1);
 
-        const bool __is_cpu = __exec.queue().get_device().is_cpu();
-        const std::uint32_t __leaf = __is_cpu ? 16 : 4;
-        _Size __steps = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __leaf);
-
+        const std::uint32_t __leaf = __leaf_sorter.__process_size;
         // 1. Perform sorting of the leaves of the merge sort tree
         sycl::event __event1 = __exec.queue().submit([&](sycl::handler& __cgh) {
             oneapi::dpl::__ranges::__require_access(__cgh, __rng);
-            __cgh.parallel_for<_LeafSortName...>(sycl::range</*dim=*/1>(__steps), [=](sycl::item</*dim=*/1> __item_id) {
-                const _IdType __i_elem = __item_id.get_linear_id() * __leaf;
-                __leaf_sort_kernel()(__rng, __i_elem, std::min<_IdType>(__i_elem + __leaf, __n), __comp);
-            });
+            __leaf_sorter.initialize_storage(__cgh);
+            const std::uint32_t __wg_count = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __leaf);
+            const sycl::nd_range<1> __nd_range(sycl::range<1>(__wg_count * __leaf_sorter.__workgroup_size),
+                                               sycl::range<1>(__leaf_sorter.__workgroup_size));
+            __cgh.parallel_for<_LeafSortName...>(__nd_range,
+                                                 [=](sycl::nd_item<1> __item) { __leaf_sorter.sort(__item); });
         });
 
         // 2. Merge sorting
         oneapi::dpl::__par_backend_hetero::__buffer<_ExecutionPolicy, _Tp> __temp_buf(__exec, __n);
         auto __temp = __temp_buf.get_buffer();
         bool __data_in_temp = false;
         _IdType __n_sorted = __leaf;
+        const bool __is_cpu = __exec.queue().get_device().is_cpu();
         const std::uint32_t __chunk = __is_cpu ? 32 : 4;
-        __steps = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __chunk);
+        const std::uint32_t __steps = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __chunk);
```

Comment:
Is `uint32_t` enough since `__n` is `size_t`?

Reply:
It may be not enough, thanks for noticing. I've corrected it.